### ISSUE TITLE
Installed but not having any response with Developer trials

### ISSUE THREAD

**Issue by tamahome23** [IDmedia/hass-nohassle_hdmi_matrix#2]
I'm currently running Home Assistant 2021.6.3 (OS) and just installed this integration through HACS.  I updated the configuration.yaml to add in the necessary names and proper IP address for my set up (I use the web interface often so know that the IP is good and the matrix is functioning to receive and process).  I do see the entity for each of my zones showing up in HASS and see the respective sources assigned for each.  That said, I've tried setting the source by several methods in HASS but it doesn't seem to be working out.  For simplicity I did the trials in developer using the  following format to no avail:

service: media_player.hdmi_matrix_set_zone
data:
  entity_id: media_player.matrix_desktv
  source: MacMini

I did the Chrome capture of the keypress as you had requested one of the previous posts to see if it helps in troubleshooting.
[NoHassleMatrix8x8-OutPut1andInput1.txt](https://github.com/IDmedia/hass-nohassle_hdmi_matrix/files/6634364/NoHassleMatrix8x8-OutPut1andInput1.txt)

I see the @T 00 00 ! command there for the press which reflects whats in the NoHassle support documents as well (also attached) so mot sure what else could be off.
[matrix_hdmi_8x8_4k_hdr_simplified.txt](https://github.com/IDmedia/hass-nohassle_hdmi_matrix/files/6634369/matrix_hdmi_8x8_4k_hdr_simplified.txt)

Any assist is much appreciated.

**Comment by IDmedia** [comment]
If you can provice me with a .har file I can look into it for you.
Here is a shot guide if you need it: https://github.com/IDmedia/hass-nohassle_hdmi_matrix/issues/1#issuecomment-853114633

The integration uses the WebUI to control the device and not UDP-packages (@t 00 00 !). Reason for that is that I wanted to be able to see the currently selected source and it was easier to get that through the WebUI. Anyways.. drop me the .har file so I can take a look at what firmware you have and take it from there. I can also add more logging in case of errors in the next update. I have no idea how many different revisions there are of this firmware/device.

**Comment by tamahome23** [comment]
Sure thing, the attachment on the first post "NoHassleMatrix8x8-OutPut1andInput1.txt" is actually the paste of the .HAR from the keypress.   Thanks for the quick response!  Let me know if I can send anything else that might help.  For now I'm using Harmony but its only purpose now is for the matrix communication, so I'll be happy to be able to shut it down completely and reclaim some cables and power if this works out!  :)

**Comment by IDmedia** [comment]
The file only contains a single POST. This is the first time I have seen "video.set" being used, but apperantly it accepts UDP-packages so it shouldn't be hard to add to this integration.

However, if you keep the DevTools window open and change the source... does it not log any other requests? Does it not show you the current selected source in the interface? I would expect to see other GET/POST requests whenever the Matrix tries to update the UI. Does for example GET/POST `http://192.1.20.68/video.get` return anything useful?

Could you please post a screenshot of what your WebUI looks like? The reason why I didn't use UDP packages for the integration is that it didn't work great on my HDMI Matrix and NoHassleAV had noe explaination why. I later figured out that my device uses different codes and since you can't get the current selection using only UDP packages I've decided to rather use the same endpoints as the web interface... I'm really dissapointed that there is no information on firmware versions as this is the third fork I've seen.

**Comment by tamahome23** [comment]
I went ahead and cleared the cache out and did the capture again, that caused it to do the heartbeats you were looking for to show up for the .HAR.  Hopefully this one will help more.  Not sure why the page I had left open didn't do the same already.  The web UI screenshot is here as well.  I'm honestly surprised to see that many more drastic variant in the handling for each of the firmwares to this point - I would think that there is little to no gain in switching it up so much on this end of things with them, but I suppose thats out of my paygrade to make those assumptions...  :)

Let me know if you would like anything else on my end.

![swbinterface](https://user-images.githubusercontent.com/42719431/121742586-a45f2900-cac5-11eb-820f-1587ddc23f60.PNG)

[NoHassleMatrix8x8-OutPut1andInput1.txt](https://github.com/IDmedia/hass-nohassle_hdmi_matrix/files/6640746/NoHassleMatrix8x8-OutPut1andInput1.txt)

**Comment by IDmedia** [comment]
Thanks for the file.

As mentioned before, setting the input should not be any issue to implement, but I'm still a bit curious about getting the current input state. Could you test opening `http://192.1.20.68/all_dat.get` in your browser to see if it always return the current state? In the .har file I see that it returns the state once, but later requests return `None`. In the .har file the request URL also appends the current date/time, but I think that's just nonesense... mine does something similar just with unix time and it still works great even if I omit it.

If that URL always yields something like the following then I'll add support for your model as well.
`@T 00 02 !@T 01 02 !@T 02 02 !@T 03 02 !@T 11 00 !@T 12 00 !@T 13 00 !@T 14 00 !@T 05 00 62 !@T 06 01 08 !@T 07 01 02 !@T 08 01 02 !@T 15 01 22 !@T 16 01 02 !@T 17 02 04 !@T 18 01 04 !`

Might take a day or two though as the other implementation I've seen has been using an array and not the "UDP-codes" or whatever they are called so have to make a parser for that. I'll update the integration to also include more logging when things can't be setup right, so I'll let you know once you can test a beta-version.

I guess there's much difference between models as NoHassleAV just rebrands the OEM HDMI Matrix.. I've seen other shops selling almost identical hardware just with another logo slapped on top. Unfortunately no one wants to reply to my firmware-related questions.

**Comment by IDmedia** [comment]
I've had some time to look at this tonight and this should work as long as I got the commands right and your Matrix always sends a reponse back when requesting the current state using `all_dat.get`.

Please try to replace the [media_player.py](https://github.com/IDmedia/hass-nohassle_hdmi_matrix/files/6642626/media_player.py.txt) file in your `config\custom_components\nohassle_hdmi_matrix` directory. Remember to remove the .txt extension and reboot your Home Assistant instanse. I'll update the code on github once this works.

Let's say you defined the following in your YAML:
```
....
    zones:
      1:
        name: Main TV Source
....        
```
Then use "Developer Tools->States" to look for `media_player.main_tv_source` and make sure the the `source` attribute it set to whatever is the current input of that output. If it is missing then the integration is struggling to get a reply/parse the current selection. Then you can see if you can change the output using the service provided in the Readme.

Please let me know how this works for you and if it doesn't what part so I can look into it :)

**Comment by tamahome23** [comment]
I don't see a source attribute showing as being reported, only the source_list off to the side.  Here is a quick screenshot of that part:

![MarixSnipping](https://user-images.githubusercontent.com/42719431/121788779-aee8f380-cb95-11eb-8361-1dab13e7d14e.PNG)

Trying the Services out in developer yields no response as well using the following:

service: media_player.hdmi_matrix_set_zone
data:
  entity_id: media_player.matrix_desktv
  source: MainAppleTV

I'll poke a few other areas to see if I can find anything else and maybe reload everything a couple more times for superstition/voodoo sake.

**Comment by IDmedia** [comment]
Thats strange... but I have a few theories. You still haven't given me an answer for what you get by visiting `http://192.1.20.68/all_dat.get` in your browser as far as I can see. This is the URL I use in order to determine wether or not your Matrix is supported and get it's current state (the source attribute). So if that doesn't return what I expect the rest won't work...

Please give this one a try ([media_player.py.txt](https://github.com/IDmedia/hass-nohassle_hdmi_matrix/files/6643528/media_player.py.txt)). Replace the file and clear your Logs before restarting Home Assistant. Right after reboot you should seee three ERRORS-entries that start with "HDMI Matrix...". Please copy and paste those three here so I can get an understanding of what's going on. One should say "API Mode" 3 and the other should give you the response from `http://192.1.20.68/all_dat.get` while the last should parse the reponse into an array of currently selected inputs (starting from 0).

Something like this:
![2021-06-13 11 13 29](https://user-images.githubusercontent.com/450724/121804589-668a0e00-cc47-11eb-9649-fc691f2af4e1.jpg)

**Comment by IDmedia** [comment]
@tamahome23  Did you have the time to test my latest changes mentioned in my last post?

**Comment by tamahome23** [comment]
Sorry bud, got caught up with a crazy amount of work - I'll give it a go tonight and follow-up.

________________________________

**Comment by tamahome23** [comment]
Just these two appear:

Unable to retrieve the HDMI Matrix state
12:03:40 PM – (ERROR) nohassle_hdmi_matrix (custom integration) - message first occurred at 12:03:40 PM and shows up 4 times
HDMI Matrix (api_mode) 1
12:03:39 PM – (ERROR) nohassle_hdmi_matrix (custom integration)

**Comment by IDmedia** [comment]
Strange, if you get API_MODE 1 that means that the integration is able to request data from: http://192.1.20.68/AutoGetAllData without errors, meaning it returns a (empty?) response or at least not an invalid one.

Could you try going to that URL in your browser and let me know what happens? I suspected it would fail with some kind of error, thus trying api_mode 2, failing again and then settle on api_mode 3... but this is the source or your problems.. so if I can improve this logic the rest should work...

**Comment by tamahome23** [comment]
It pulls up a blank page that just has one word:
none

**Comment by IDmedia** [comment]
That explains it, give me 2 minutes

**Comment by IDmedia** [comment]
Please give this one a try. The added check makes sure that the Matrix returns a valid array of inputs when configuring the integration. This way it should log api_mode 3 (hopefully)...

I'll check back in an hour or so, would be nice to get this issue resolved too :)

[media_player.py.txt](https://github.com/IDmedia/hass-nohassle_hdmi_matrix/files/6682786/media_player.py.txt)

**Comment by tamahome23** [comment]
nohassle_hdmi_matrix: Error on device update!
12:21:41 PM – (ERROR) Media player (custom integration) - message first occurred at 12:21:41 PM and shows up 4 times
Error connecting to the HDMI Matrix
12:21:40 PM – (ERROR) nohassle_hdmi_matrix (custom integration)
HDMI Matrix (api_mode) None
12:21:40 PM – (ERROR) nohassle_hdmi_matrix (custom integration)

**Comment by IDmedia** [comment]
Okay... Looks like it's still not getting the correct information from the Matrix...

Please try again with [this file](https://github.com/IDmedia/hass-nohassle_hdmi_matrix/files/6682849/media_player.py.txt). A log entry called [http://192.168.1.168/all_dat.get
](http://192.1.20.68/all_dat.get) will be generated after a reboot. Please copy the whole message of that error here (click on it so you get the details).

I suspect that `http://192.1.20.68/all_dat.get` isn't responding with the current input state. Please also open it in your browser and let me know that you get a response.

The integration expects a response similar to this:
`@T 00 02 !@T 01 02 !@T 02 02 !@T 03 02 !@T 11 00 !@T 12 00 !@T 13 00 !@T 14 00 !@T 05 00 62 !@T 06 01 08 !@T 07 01 02 !@T 08 01 02 !@T 15 01 22 !@T 16 01 02 !@T 17 02 04 !@T 18 01 04 !`

**Comment by tamahome23** [comment]
nohassle_hdmi_matrix: Error on device update!
10:40:26 PM – (ERROR) Media player (custom integration) - message first occurred at 10:40:26 PM and shows up 4 times
Error connecting to the HDMI Matrix
10:40:25 PM – (ERROR) nohassle_hdmi_matrix (custom integration)
http://192.1.20.68/all_dat.get
10:40:25 PM – (ERROR) nohassle_hdmi_matrix (custom integration)

**Comment by tamahome23** [comment]
And the detail is:

Logger: custom_components.nohassle_hdmi_matrix.media_player
Source: custom_components/nohassle_hdmi_matrix/media_player.py:97
Integration: nohassle_hdmi_matrix (documentation)
First occurred: 10:40:25 PM (1 occurrences)
Last logged: 10:40:25 PM

http://192.1.20.68/all_dat.get
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/urllib/request.py", line 1354, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/usr/local/lib/python3.8/http/client.py", line 1252, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/local/lib/python3.8/http/client.py", line 1298, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.8/http/client.py", line 1247, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/local/lib/python3.8/http/client.py", line 1007, in _send_output
    self.send(msg)
  File "/usr/local/lib/python3.8/http/client.py", line 947, in send
    self.connect()
  File "/usr/local/lib/python3.8/http/client.py", line 918, in connect
    self.sock = self._create_connection(
  File "/usr/local/lib/python3.8/socket.py", line 808, in create_connection
    raise err
  File "/usr/local/lib/python3.8/socket.py", line 796, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/nohassle_hdmi_matrix/media_player.py", line 97, in setup_platform
    with urllib.request.urlopen(f'http://{host}/all_dat.get', timeout=5) as response:
  File "/usr/local/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/local/lib/python3.8/urllib/request.py", line 525, in open
    response = self._open(req, data)
  File "/usr/local/lib/python3.8/urllib/request.py", line 542, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/usr/local/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/local/lib/python3.8/urllib/request.py", line 1383, in http_open
    return self.do_open(http.client.HTTPConnection, req)
  File "/usr/local/lib/python3.8/urllib/request.py", line 1357, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [Errno 111] Connection refused>

**Comment by tamahome23** [comment]
As for the last question, when I perform the 'http://192.1.20.68/all_dat.get' in a browser it comes back with 'none' like on the other.

**Comment by IDmedia** [comment]
Well... okay, then the URL is incorrect or the Matrix is not behaving right...
What about `http://192.1.20.68/all_dat.get0` or `http://192.1.20.68/all_dat.getFri%20Jun%2011%202021%2014:48:33%20GMT-0500%20(Central%20Daylight%20Time)`?

Not sure if the last one requires a date in the future, but please give that one a try....

**Comment by tamahome23** [comment]
Both return 'none'

**Comment by IDmedia** [comment]
So when you request `http://192.1.20.68/all_dat.get` it returns just `None`.
The only other suggestion I have is to try to hit `http://192.1.20.68/ip.get` AND THEN immediately afterwards `http://192.1.20.68/all_dat.get` to see if that changes anything. You might want to close any window you have open to the WebUI to make sure that doesn't somehow interfere with the testing.

I see that only two requests to `all_dat.get` which return the correct data in your .har file and both of those are triggered AFTER requesting `ip.get` or changing inputs.

I find it strange if the Matrix only gives you the status after a change, but well.. if that doesn't work either then I'm out of ideas for now given that I can't test the hardware myself directly.

**Comment by tamahome23** [comment]
The ip.get returns:
settingsCallback({"ver":"1.0", "mac":"00:08:DC:FE:CB:6E", "ip":"192.1.20.68", "gw":"192.1.20.1", "sub":"255.255.255.0", "dhcp":"1", });

while the all_dat.get returns:
none

If it would help I could coordinate a teams/zoom call or something to share control during a video call - outside of that not much else I could think of either.  I'm still ok for now myself as the harmony is running the matrix, but as you know there really is no reason that the web calls shouldn't work and do so infinitely better (no worry about signal being blocked by kid or object).

Ah well, if you do come up with something you want me to try out further, feel free to reach out and let me know.

**Comment by ripolltata** [comment]
Hello, good afternoon!
I had previously consulted you because the integration did not work well. I finished solving it with a universal ir. But now I would like to know the status of the matrix to show it in some buttons, so I went back to investigate and discovered that you had made an update. Great! I updated and I started to test it. I was very happy to see Developer Tools and now if I saw the information in attributes.
The bad news was that he only showed it to me the first time and never again
now I'm the same as before ...
I'm going to send you some screenshots to see if they work for you
I am working with home assitant core-2021.3.4
my matrix es 8ch 4K
I could also copy the har file to inspect it, but I can't find how to share it

Thank you very very much

![matrix_atributos](https://user-images.githubusercontent.com/77713471/123525271-c354e200-d6a5-11eb-8828-77847aa2ff95.png)
![Matrix01](https://user-images.githubusercontent.com/77713471/123525326-1c247a80-d6a6-11eb-9774-4c13fe8ac753.png)

![Matrix02](https://user-images.githubusercontent.com/77713471/123525278-cea80d80-d6a5-11eb-9be6-ec248cd66537.png)

**Comment by IDmedia** [comment]
@ripolltata It would be better if you had created a new issue as this one is not related to your issues. You seem to have a similar UI to what was discussed in this [issue](https://github.com/IDmedia/hass-nohassle_hdmi_matrix/issues/1). There seem to be a lot of variations of the WebUI, so I can't promise the integration will work for everyone without getting the .har file. If you have trouble sharing it, you can always rename it as a .txt file by opening it in Notepad (or similar) and save it again with a .txt extention.

Anyways.. the missing `Source` attribute tells me that the integration fails to get the current state of the Matrix. Please make sure you are using the latest version, clear the logs and restart HA. If the setup fails you should see `Error connecting to the HDMI Matrix` in the logs now.

If you get that error I need the .har file and information about what you get when visiting this URL in your browser: `http://{IP_OF_YOUR_MATRIX/cgi-bin/query` in order to debug the issue.

**Comment by ripolltata** [comment]
Hi IDmedia, Thanks for replying ...
I do not have much experience in this and consult here because it seemed to me that the subject was related.
no problem, if you want I can open a new issue.
I'm going to do what you recommended and I'll tell you anything.
Thank you

**Comment by IDmedia** [comment]
@tamahome23 When looking at the other .har files I see that the Matrix doesn't always respond with the correct status unless you make another request first. I'm not sure, but I wonder if that's the same on yours.. so I modified the integration to try to always request `ip.get` before getting the state. Care to give that one a try?

[media_player.py.txt](https://github.com/IDmedia/hass-nohassle_hdmi_matrix/files/6732119/media_player.py.txt)

If that also doesn't work we'll have to look into finding other commands that might give the status every time and not just `None`, but I haven't found anything new just yet.

**Comment by tamahome23** [comment]
No love I'm afraid - didn't seem to act any differently.

________________________________

**Comment by IDmedia** [comment]
```
function httpGetDate()
{
	var xmlHttp = new XMLHttpRequest();
	xmlHttp.open( 'GET', 'all_dat.get'+new Date(), true );
	xmlHttp.send(); 
	xmlHttp.onreadystatechange = function(){
    if (xmlHttp.readyState == 4 && xmlHttp.status == 200){
			SynRadio(xmlHttp.responseText);
			setTimeout('httpGetDate()',2000);
    }
  }	  
}
```

This seems to be the code handling the `all_dat.get` requests. There is no other logic then parsing the data if when it's available, so I don't quite understand why this isn't working as expected... There must be some stupid request/parameter/header missing, but the only difference I see between the requests with a valid response and not is the Content-Length (obviously)
